### ISSUE TITLE
Holes

### ISSUE THREAD

**Issue by luc4leone** [arosenfeld2003/testingJavascriptArrayMethods#3]
Hi @arosenfeld2003 

I think you didn't take into account holes (sparse arrays).

Compare and contrast the result of these 2 statements:

```javascript
[1, 2, 3, 4, 5, 6, , 7].copyWithin(7, 6);
[1, 2, 3, 4, 5, 6, , 7].myCopyWithin(7, 6);
```

**Comment by arosenfeld2003** [comment]
Luca,
This was another interesting discovery when experimenting with the example you give above:

The console logs the last two values as 'empty' on the native method.  
On my method, they are logged as 'empty', followed by 'undefined'.

I think this is simply a quirk of the Chrome debugger.  The actual value of the 'empty' index logs to the console as 'undefined', so I am satisfied with my code and tests for holes.

Try this in the console:

var arr1 = [1, 2, 3, 4, 5, 6, , 7].copyWithin(7, 6);
var arr2 = [1, 2, 3, 4, 5, 6, , 7].myCopyWithin(7, 6);

for (var i = 0; i < myCopied.length; i++) {
	console.log(myCopied[i] === copied[i]);
}

The result is true for every index in the array.

Thanks for your feedback: this has also been a good learning exercise.

**Comment by luc4leone** [comment]
>I think this is simply a quirk of the Chrome debugger.

I don't think so @arosenfeld2003 . When you run `var arr2 = [1, 2, 3, 4, 5, 6, , 7].myCopyWithin(7, 6);` index 7 is there and it shouldn't be.

<img src="https://user-images.githubusercontent.com/2797254/53224894-8c560800-3676-11e9-8794-58a707a0f8e8.png" width="400" />

**Comment by arosenfeld2003** [comment]
Right ... but if you check the value of index 7 of the native method, you get ‘undefined’ right?

Sent from my iPhone

Alex Rosenfeld
215-301-5343

**Comment by luc4leone** [comment]
>Right ... but if you check the value of index 7 of the native method, you get ‘undefined’ right?

Right. But one thing is to have an element with `undefined` value, another thing is to not have the element, or did I misunderstand what you're implying?

**Comment by arosenfeld2003** [comment]
Luca,

You are correct.  Another mis-assumption on my part.  It is interesting because the length of the new array stays the same.  Also, after more detailed look at my code, I realized I was not testing for this properly... another refactor in order!

Thanks for your suggestions.

I’d be curious about what strategy you used to copy ‘holes’ without adding an ‘undefined’ element.

I found this website which seems pretty detailed (I like the use of a chart/table… this is definitely ‘Gordon-Approved’ haha).

http://2ality.com/2015/09/holes-arrays-es6.html <http://2ality.com/2015/09/holes-arrays-es6.html>

Cheers,
Alex

**Comment by luc4leone** [comment]
>I found this website which seems pretty detailed

For holes I have found this link useful http://speakingjs.com/es5/ch18.html... we are using the same ref basically, since it's the same guy :)

>I’d be curious about what strategy you used to copy ‘holes’ without adding an ‘undefined’ element.

you can check my strategy here (not sure I like it yet): 
https://gist.github.com/luc4leone/70f09ad4636b37ae7a03b20c2ff88cc3

>Thanks for your suggestions.

Thank you Alex for the nice conversation :)

**Comment by arosenfeld2003** [comment]
Thanks for sharing your code.  We’ve obviously approached this quite differently.  I really enjoy that aspect of programming - seeing how different solutions can solve the same problem.

I spent a while today thinking about this, experimented a lot.  Finally, I realized how to copy the holes into the new array.

Feel free to have a look:

https://github.com/arosenfeld2003/testingJavascriptArrayMethods/blob/master/myCopyWithin.html <https://github.com/arosenfeld2003/testingJavascriptArrayMethods/blob/master/myCopyWithin.html>

Cheers,
Alex
